adding pipenv support

elreydetoda opened the pull request:
Thank you for this contribution! While pipenv was a good solution when this PR was created in 2020, the Python packaging ecosystem has evolved significantly since then.

We're closing this PR as the community has largely moved beyond pipenv to more modern solutions. We'll be exploring simpler packaging options that minimize setup requirements for non-developer users.

We appreciate your effort in improving Algo's developer experience!

Fuckmylimpdick commented:
No

elreydetoda commented:
> No

Ok, @Fuckmylimpdick is there any reason in particular as to why you are saying no?

jauderho commented:
+1 to this.

We can generate a matching requirements.txt from the Pipefile for folks that want to continue using that but in general, I've found that pipenv has been better at working out the constraints & dependencies.

elreydetoda commented:
It's been a bit since I have installed the deps for that branch, but I should easily be able to update this PR if someone from the project wants to accept it :slightly_smiling_face:

jauderho commented:
I'd say to wait after #14272 is merged

elreydetoda commented:
tl;dr: this has the exact same dependencies that the requirements.txt has in master currently.

So, this Pipfile & Pipfile.lock should be pinning to the exact dependencies inside the current [requirements.txt](https://github.com/trailofbits/algo/blob/4bed66f19eca9636528985cac5f6a725bc2c8739/requirements.txt).

I added an [intermediary commit](https://github.com/trailofbits/algo/pull/14114/commits/aa61ffcc8668ec4c3c7d1cdeed5135ee4b497a5a) in there, so I could add `"*"` for all the dependencies.  This will make upgrades in the future easier, but in the last commit you can see I reverted the hashes from the initial commit and updated the `hash` line specifically so Pipenv doesn't try to update the lock file again.


```console
$ gd HEAD..5210bf4 | cat
diff --git a/Pipfile b/Pipfile
index 3926440..8002695 100644
--- a/Pipfile
+++ b/Pipfile
@@ -4,10 +4,10 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-ansible-core = "*"
-ansible = "*"
+ansible-core = "==2.11.3"
+ansible = "==4.4.0"
 netaddr = "*"
-Jinja2 = "*"
+Jinja2 = "~=3.0.1"
 
 [dev-packages]
 
diff --git a/Pipfile.lock b/Pipfile.lock
index d5b4e73..bed9528 100644
--- a/Pipfile.lock
+++ b/Pipfile.lock
@@ -1,7 +1,7 @@
 {
     "_meta": {
         "hash": {
-            "sha256": "0c3c627d43a80cad0a384a195031ecdf469cea355e4b4e878b0a5f089e925d25"
+            "sha256": "1277a5279f6bb88863ea7a08d9bf546dca26dab2c405640d45fc5b5f958f9ec8"
         },
         "pipfile-spec": 6,
         "requires": {
```


Speaking of updates, the way you should be able to update dependencies with pipenv is `pipenv update` command.

I say *should*, because ( like @glennschler had an issue with [here](https://github.com/trailofbits/algo/pull/14272#issuecomment-958134347) ) sometimes ansible wants a fresh install.  So, in that can I have had to run `pipenv --rm && pipenv install --skip-lock && pipenv lock`.  This will essentially install all packages "fresh" and then pin the dependencies after the install.
